lib/metasploit/framework/credential_collection.rb

```diff
@@ -219,12 +233,119 @@ def add_public(public_str='')
       additional_publics << public_str
     end
 
-    # Combines all the provided credential sources into a stream of {Credential}
-    # objects, yielding them one at a time
-    #
-    # @yieldparam credential [Metasploit::Framework::Credential]
-    # @return [void]
-    def each_unfiltered
+    # When password spraying is enabled, do first passwords then userames
+    #  i.e.
+    #   username1:password1
+    #   username2:password1
+    #   username3:password1
+    # ...
+    #   username1:password2
+    #   username2:password2
+    #   username3:password2
+    # ...
+    def each_unfiltered_password_first
+      if user_file.present?
+        user_fd = File.open(user_file, 'r:binary')
```

Comment:
I didn't verify locally; but from an eyeball - I believe your test failure is caused by expecting a `proc` to be provided here:

```ruby
      let(:user_file) do
        filename = "user_file"
        stub_file = StringIO.new("user1\nuser2\nuser3\n")
        allow(File).to receive(:open).with(filename,/^r/).and_yield stub_file

        filename
      end
```

but you're using `File.open(...)` without a proc, unlike (I imagine) the original implementation/tests

Reply:
I copied:
```
context "when given a pass_file and user_file" do
      let(:password) { nil }
      let(:username) { nil }
      let(:user_file) do
        filename = "user_file"
        stub_file = StringIO.new("asdf\njkl\n")
        allow(File).to receive(:open).with(filename,/^r/).and_yield stub_file

        filename
      end
      let(:pass_file) do
        filename = "pass_file"
        stub_file = StringIO.new("asdf\njkl\n")
        allow(File).to receive(:open).with(filename,/^r/).and_return stub_file

        filename
      end

      specify  do
        expect { |b| collection.each(&b) }.to yield_successive_args(
          Metasploit::Framework::Credential.new(public: "asdf", private: "asdf"),
          Metasploit::Framework::Credential.new(public: "asdf", private: "jkl"),
          Metasploit::Framework::Credential.new(public: "jkl", private: "asdf"),
          Metasploit::Framework::Credential.new(public: "jkl", private: "jkl"),
        )
      end
    end
```

And just added a param and made the user and pass more understandable

Reply:
I believe your implementation differs, as you are doing:

```
        user_fd = File.open(user_file, 'r:binary')
```

Instead of:

```
       File.open(user_file, 'r:binary') do |user_fd|
          # ... do things with user_fd ...
       end
```

Therefore this mock fails:

```
        filename = "user_file"
        stub_file = StringIO.new("asdf\njkl\n")
        allow(File).to receive(:open).with(filename,/^r/).and_yield stub_file
                                                          ^^^^^^^^^^^^^^^^^^^
```

So I think you'll want to either align the implementation with the test's expectations, or update the test to change its expectations - i.e. probably using `allow(File).to receive(:open).with(filename,/^r/).and_return stub_file` instead of `.and_yield stub_field`

Reply:
Didn't help :(

Unfortunately my primary dev language is Python and not Ruby - so I struggling with this

I cannot change the way the credentials_collection.rb works, it has to open the user file at the beginning in order to be able to yield combinations of the same password and different usernames